_posts/2020-12-25-manual-gates-git-flow-in-a-wig.md

```diff
@@ -0,0 +1,122 @@
+
+<center><img src="/img/manual-gates/simple-1.png" /></center>
+<br/>
+Most teams will start with this. At some point, the team itself will decide there is too much risk deploying directly to production, and will decide they need an additional environment. Enter the test (or nonproduction, or development, or QC) stage. They put in a manual gate between deployment phases and everything is fine. The pipeline starts to get used in other projects and the world keeps turning.
+
+<center><img src="/img/manual-gates/simple-2.png" /></center>
+<br/>
+Then a new project comes along, and decides two stages isn't enough. They want another one. It could be for User Acceptance Testing. It could be for training. At any rate it gets shoved into the pipeline between the nonproduction and production stages and the world kicks on; although the teams have noticed that this project does not deploy to production as often.
+
+<center><img src="/img/manual-gates/simple-3.png" /></center>
+<br/>
+More time passes. Other teams decide they need another environment. Others add other types of verification steps into the pipeline. As they get more and more complicated, the likelihood of code getting through all the way without manual intervention decreases. Frustrated with increasing complexity, teams find ways to circument the pipeline in order to deploy manually. Incidents start to increase in frequency and duration. Engineers moving between teams are frustrated because the pipelines have wildly diverged.
+
+Sounding familiar?
+
+The purpose of your software is to create value for your users - the purpose of your pipeline is to ensure users can get timely access to your software. The faster you can push working functionality through your pipeline, the more value you can create for your users. Building your software delivery pipeline around this core tenent is critical to value capture. The addition of time-consuming and/or manual processes between commit (or merge) and deployment is responsible for a significant amount of drag.
+
+> First of all, nobody cares how cool your CI/CD pipeline is. - Forrest Brazeal
+
+I've watched this story unfold multiple times. I've noticed most commonly that it arises out of a need for two things - manual gating and multiple environments. On the surface both of these things are entirely reasonable options but they seem to result in a productivity death spiral. I've since come to the conclusion that these are tools that one needs to have in their toolbox - the problem stems from how teams choose to implement them. I believe that these two features should be implemented via your branching model and not via discrete steps in the CI/CD pipeline. So what branching model should you use? Let's first consider the GitFlow branching model.
```

Comment:
I'm voting for a touch more alarm-ism to grab people
The thinking is that both of these things are prudent safety measures, more frequently however it's a gateway drug to a shaky deployment process and a productivity death spiral.